### PR TITLE
understand capture operator in Elixir

### PR DESCRIPTION
**Pull request by Daniel-Xu**
Daniel
for 08-05
[Rendered](https://github.com/DockYard/reefpoints/blob/2016-08-05-understand-capture-operator-in-elixir/source/posts/2016-08-05-understand-capture-operator-in-elixir.md)

### CODE REVIEW

**Comment by bcardarella** [comment]
Nice post :)

If you wanted to show something crazy you can with capture, but something that should be a *don't ever do this*

```elixir
foo
|> &(bar(baz, &1))
```

(syntax may be a little off)

People are always asking "how do I pipe into the Nth argument instead of the first". The correct answer is "you can't".... but that's not 100% true. You can hack your way around this with the capture operator. What happens is that you can create a new anonymous function that you're piping into that takes the one argument and then use that on whatever argument order you'd like inside the anon function.

Once again, don't ever do this.

**Comment by Daniel-Xu** [comment]
@bcardarella  ha, good to know that! I feel more comfortable reading it after writing this post.

File: source/posts/2016-08-05-understand-capture-operator-in-elixir.md [hunk 1]
```diff
@@ -0,0 +1,99 @@
+---
+layout: post
+title: "Understanding & (capture operator) in Elixir"
+summary: "how to use & (capture operator) in Elixir"
+author: "Daniel Xu"
+twitter: "Daniel_Xu_For"
+github: "Daniel-Xu"
+published: true
+tags: capture operator, Elixir
```
Comment (rsocci) [inline source/posts/2016-08-05-understand-capture-operator-in-elixir.md]: `capture operator` tag not needed as it is too specific

File: source/posts/2016-08-05-understand-capture-operator-in-elixir.md [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+---
+
+# Understanding & (capture operator) in Elixir
+
+`&` is capture operator in Elixir, it is used to **capture** and **create** anonymous function.
```
Comment (rsocci) [inline source/posts/2016-08-05-understand-capture-operator-in-elixir.md]: function -> function`s`

File: source/posts/2016-08-05-understand-capture-operator-in-elixir.md [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+
+## Anonymous function and arity
```
Comment (rsocci) [inline source/posts/2016-08-05-understand-capture-operator-in-elixir.md]: functions

File: source/posts/2016-08-05-understand-capture-operator-in-elixir.md [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+
+Before going into details about capture operator, let's get familiar with `anonymous function` and `arity` first.
```
Comment (rsocci) [inline source/posts/2016-08-05-understand-capture-operator-in-elixir.md]: about **the** capture operator

Comment (rsocci) [inline source/posts/2016-08-05-understand-capture-operator-in-elixir.md]: function*s*

File: source/posts/2016-08-05-understand-capture-operator-in-elixir.md [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+
+Given the following example:
+
+```elixir
+add_one = fn x -> x + 1 end
+
```
Comment (rsocci) [inline source/posts/2016-08-05-understand-capture-operator-in-elixir.md]: extra line

File: source/posts/2016-08-05-understand-capture-operator-in-elixir.md [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+```
+we defined a function, but it isn't bound to a global name, so it is an anonymous function or a lambda.
```
Comment (rsocci) [inline source/posts/2016-08-05-understand-capture-operator-in-elixir.md]: capitalize `We`

File: source/posts/2016-08-05-understand-capture-operator-in-elixir.md [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+we capture `puts` function from `IO` module and bind it with a local name `speak`.
+
+* local function
+
+In the following example, `put_in_columns` and `put_in_one_row` are defined in the same module, so we can capture `put_in_one_row` by
+`&put_in_one_row/1`, notice that we don't include the module name here.
+
+```elixir
+defmodule Issues.TableFormatter do
+
```
Comment (rsocci) [inline source/posts/2016-08-05-understand-capture-operator-in-elixir.md]: extra line here and at the end of the example

File: source/posts/2016-08-05-understand-capture-operator-in-elixir.md [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+author: "Daniel Xu"
+twitter: "Daniel_Xu_For"
+github: "Daniel-Xu"
+published: true
+tags: Elixir
+---
+
+# Understanding & (capture operator) in Elixir
+
+`&` is capture operator in Elixir, it is used to **capture** and **create** anonymous functions.
```
Comment (duggiefresh) [inline source/posts/2016-08-05-understand-capture-operator-in-elixir.md]: `&` is the capture operator in Elixir...

File: source/posts/2016-08-05-understand-capture-operator-in-elixir.md [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+title: "Understanding & (capture operator) in Elixir"
+summary: "how to use & (capture operator) in Elixir"
+author: "Daniel Xu"
+twitter: "Daniel_Xu_For"
+github: "Daniel-Xu"
+published: true
+tags: Elixir
+---
+
+# Understanding & (capture operator) in Elixir
```
Comment (duggiefresh) [inline source/posts/2016-08-05-understand-capture-operator-in-elixir.md]: ... (the capture operator) in Elixir

**Comment by duggiefresh** [comment]
few comments, LGTM

File: source/posts/2016-08-05-understand-capture-operator-in-elixir.md [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+```
+
+is the same with:
+
+```elixir
+add_one = fn x -> x + 1 end
+add_one.(1) # 2
+```
+
+you might notice that there's `&1` in our example, that is `value placeholder`, it identifies the `nth` argument of the function.
```
Comment (rsocci) [inline source/posts/2016-08-05-understand-capture-operator-in-elixir.md]: Reword suggestion: "You might notice that`&1` is used in the above example. That's called a `value placeholder`, and it identifies the `nth` argument of the function.

File: source/posts/2016-08-05-understand-capture-operator-in-elixir.md [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+
+In addition, as `{}` and `[]` are also operator in Elixir, so `&` can work with them too.
```
Comment (rsocci) [inline source/posts/2016-08-05-understand-capture-operator-in-elixir.md]: operator -> operators

Comment (rsocci) [inline source/posts/2016-08-05-understand-capture-operator-in-elixir.md]: can remove word `so`

File: source/posts/2016-08-05-understand-capture-operator-in-elixir.md [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+return_list = &[&1, &2]
+return_list.(1, 2) # [1, 2]
+
+return_tuple = &{&1, &2}
+return_tuple.(1, 2) # {1, 2}
+```
+
+It's hard to comprehend at first, we just need to think about it from another perpective:
+
+![Alt text](https://monosnap.com/file/RfDoLHTqzOzGGXetAoUfUvVwPpAf5j.png)
```
Comment (rsocci) [inline source/posts/2016-08-05-understand-capture-operator-in-elixir.md]: nice diagram :)

**Comment by rsocci** [comment]
Great post @Daniel-Xu!!

**Comment by Daniel-Xu** [comment]
thx for the review @rsocci  @duggiefresh

**Pull request merged**